### PR TITLE
test: Move My Expenses To My Platform: Fixing Types and Errors

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 669200a</samp>

This pull request refactors and improves the custom fields logic for expenses and other models by using a new `CustomFields` interface. It also fixes some bugs and adds more test cases for various components and services that use custom fields. The affected files are `expense.model.ts`, `merge-expense-3.page.spec.ts`, `my-expenses.page.spec.ts`, `my-expenses.service.spec.ts`, `create-new-report.component.spec.ts`, and `custom-fields.model.ts`.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 669200a</samp>

> _Sing, O Muse, of the skillful coder who refines the fields of data_
> _With a new interface, `CustomFields`, that adapts to various forms and types_
> _And who tests the merging of expenses with mock inputs and assertions_
> _And who fixes bugs and filters in the specs of reports and elements_

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 669200a</samp>

*  Add a new interface for custom fields in `custom-fields.model.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-7c6af413efb2cc8145023366ea84269410e14f4e41f9d0d52d2829e41140bed0R1-R6))
*  Use the new custom fields interface in the expense model and replace the old NameValuePair type in `expense.model.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-57a2d67f855d51e1a242e5077eea76af56c28517c98ae770f47d197b126946b8R18), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-57a2d67f855d51e1a242e5077eea76af56c28517c98ae770f47d197b126946b8L38-R39))
*  Use the mock data for merge expenses custom inputs and test the merge expense page logic in `merge-expense-3.page.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bR77), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL303-R304))
*  Fix linting errors and remove trailing commas in `merge-expense-3.page.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL130-R131), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL164-R165), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL255-R256), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL266-R267), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL415-R416), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL434-R435), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL453-R454), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-a7ed545db35aa672503bbcc4009804e9fc36c5213f1041c026c657a65eaf023bL497-R498))
*  Use the mock data for expenses and test the my expenses page logic in `my-expenses.page.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7R115), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L686-R687), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L748-R749), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L759-R760), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1018-R1019), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1024-R1025), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1714-R1715), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1723-R1726), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1739-R1744), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1754-R1757), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1767-R1773), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1782-R1786), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1797-R1798), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1809-R1810), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1819-R1820), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1829-R1830), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1902-R1904), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1919-R1920), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1942-R1943), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1966-R1967), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1990-R1991), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2012-R2013), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2037-R2038), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2066-R2067), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2092-R2093), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2118-R2119), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2156-R2157), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2380-R2381), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2476-R2482), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2487-R2488), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2502-R2509), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2587-R2589), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2613-R2615), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2651-R2652), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2675-R2676), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L2798-R2799))
*  Refactor the filters logic and use the new keys and functions for the filters in `my-expenses.page.spec.ts` and `my-expenses.service.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L845-R849), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L856-R857), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L870-R875), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L881-R882), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L897-R901), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L907-R908), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1601-R1602), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1624-R1625), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-446039c75c9b2d02f68c2a029282002588ffa48a5d20add67bfbce3c26dacbb7L1645-R1646), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-ea9713753a532f13bfbc1c05db95ce32cb8774bd3e6cfb94b310a68636d3ed51L73-R73), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-ea9713753a532f13bfbc1c05db95ce32cb8774bd3e6cfb94b310a68636d3ed51L80-R80), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-ea9713753a532f13bfbc1c05db95ce32cb8774bd3e6cfb94b310a68636d3ed51L87-R87), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-ea9713753a532f13bfbc1c05db95ce32cb8774bd3e6cfb94b310a68636d3ed51L94-R94))
*  Use the mock data for expenses and test the create new report component logic in `create-new-report.component.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58R30), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58L91-R93), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58L104-R105), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58L114-R115), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58L156-R161), [link](https://github.com/fylein/fyle-mobile-app/pull/2586/files?diff=unified&w=0#diff-69140b7af7e715ffac01f5e45e47e0b5f8f93b6b19b8c9b25607192a58f69c58L166-R174))

## Clickup
https://app.clickup.com/t/8678yrj24

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/core/models/platform/custom-fields.model.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export interface CustomFields {
+  name: string;
+  value: string | string[] | boolean | number | Record<string, string | string[] | boolean>;
+  type?: string;
+  is_enabled?: boolean;
```
Comment (OmkarJ13) [inline src/app/core/models/platform/custom-fields.model.ts:5]: API will always return this, can remove the optional ig.
Let's add this to the existing mock data as well

Comment (OmkarJ13) [reply src/app/core/models/platform/custom-fields.model.ts:5]: Also pls move this model under `platform/v1/` directory.

Comment (OmkarJ13) [reply src/app/core/models/platform/custom-fields.model.ts:5]: Can you also check if `type` can be made into an enum?

**Review by OmkarJ13 — CHANGES_REQUESTED** [review]
Some comments

**Review by jayfyle — COMMENTED** [review]
Please ignore the size of the PR, added some mock data causing the size to increase drastically 🙏

File: src/app/fyle/my-expenses/my-expenses.service.spec.ts [hunk 2]
```diff
@@ -70,28 +70,28 @@ describe('MyExpensesService', () => {
       spyOn(myExpensesService, 'convertSelectedSortFitlersToFilters');
       const sortBy = { name: 'Sort By', value: 'dateNewToOld' };
 
-      const convertedFilters = myExpensesService.convertFilters(selectedFilters7);
+      // const convertedFilters = myExpensesService.getFilters(selectedFilters7);
```
Comment (Dimple16) [inline src/app/fyle/my-expenses/my-expenses.service.spec.ts:73]: why are we commenting this? if we are taking care of it in the next PR, please add a `TODO`

**Review by Dimple16 — APPROVED** [review]
added a comment

File: src/app/fyle/my-expenses/my-expenses.service.spec.ts (hunk 2 above)
Comment (jayfyle) [reply src/app/fyle/my-expenses/my-expenses.service.spec.ts:73]: Yes fixed in next PR

File: src/app/core/models/platform/custom-fields.model.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { CustomFieldTypes } from '../../enums/platform/v1/custom-fields-type.enum';
+
+export interface CustomFields {
+  name: string;
+  value: string | string[] | boolean | number | Record<string, string | string[] | boolean>;
```
Comment (Chethan-Fyle) [inline src/app/core/models/platform/custom-fields.model.ts:5]: What about date type?

**Review by Chethan-Fyle — APPROVED** [review]

**Review by OmkarJ13 — APPROVED** [review]

**Pull request merged**